### PR TITLE
feat: project location must match jurisdiction

### PR DESCRIPTION
**Pull request by ryanchristo**
## Description

Closes: regen-network/regen-registry#1499

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] provided a link to the relevant issue or specification
- [ ] provided instructions on how to test
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### How to test

https://deploy-preview-2090--regen-marketplace.netlify.app/

1. Log in with account that is a project admin for an existing project on chain
2. Go to user portfolio and edit project that exists on chain
3. Attempt to update the project location and note must match jurisdiction error

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**._

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

File: web-marketplace/src/components/organisms/LocationForm/LocationForm.tsx [hunk 1]
```diff
@@ -92,12 +94,27 @@ const LocationForm: React.FC<LocationFormProps> = ({
...
-          }
+            });
+            // If a jurisdiction is provided (i.e. the project exists on chain),
+            // we require the location input to match the jurisdiction.
+            if (jurisdiction) {
+              let valid: boolean;
+              if (isGeocodingFeature(location)) {
+                valid = location.properties.short_code === jurisdiction;
+              } else {
+                valid = location === jurisdiction;
```
Comment (blushi) [inline web-marketplace/src/components/organisms/LocationForm/LocationForm.tsx:109]: I believe we shouldn't show any error message before the user selects a location from the search results.

**Review by blushi — CHANGES_REQUESTED** [review]
I'm getting the error "This location must match the on chain jurisdiction" even when selecting a location that matches the jurisdiction. But then if I click inside the input field, it gets valid again, even if the location doesn't match.

File: web-marketplace/src/components/organisms/LocationForm/LocationForm.tsx [hunk 2]
```diff
@@ -64,10 +67,20 @@ const LocationForm: React.FC<LocationFormProps> = ({
...
+
   useEffect(() => {
     isDirtyRef.current = isDirty;
   }, [isDirtyRef, isDirty]);
 
+  useEffect(() => {
+    // If a jurisdiction is provided (i.e. the project exists on chain),
+    // we require the location input to match the jurisdiction.
+    setJurisdictionError(validateLocation(jurisdiction, location));
+  }, [jurisdiction, location]);
```
Comment (ryanchristo) [inline web-marketplace/src/components/organisms/LocationForm/LocationForm.tsx:82]: In order to get this working I had to check the `location` on render and the `value` on change within the form. I'm not sure if there is a better way of doing this and am a bit confused about how to add custom validation to react hook forms but this should be working as expected now. Let me know if I am overlooking something.

Comment (ryanchristo) [reply web-marketplace/src/components/organisms/LocationForm/LocationForm.tsx:82]: If easier done than said, happy to have someone take this over.

Comment (flagrede) [reply web-marketplace/src/components/organisms/LocationForm/LocationForm.tsx:80]: I think we could have that in the zod schema validation by using `refine` and passing the onChain jurisdiction when instantiating the schema. You can look at the [metadata schema](https://github.com/regen-network/regen-web/blob/dev/web-marketplace/src/components/organisms/MetadataForm/MetadataForm.schema.ts) for an example of how to do that.

**Comment by ryanchristo** [comment]
I'm closing this pull request and unassigning myself from the issue. I agree with the use of zod for validation but am less familiar with the tools and probably makes sense for someone else to pick this up and start from scratch.

**Pull request closed without merging**